src/charm.py

```diff
@@ -188,9 +254,91 @@ def _vault_service_is_running(self) -> bool:
+    ) -> bool:
+        """Check if the certificates are pushed to the workload."""
+        if not self._container.exists(path=TLS_CERT_FILE_PATH):
+            return False
+        if not self._container.exists(path=TLS_KEY_FILE_PATH):
+            return False
+        if not self._container.exists(path=TLS_CA_FILE_PATH):
+            return False
+        existing_certificate = self._container.pull(path=TLS_CERT_FILE_PATH)
+        if existing_certificate.read() != certificate:
+            return False
+        existing_private_key = self._container.pull(path=TLS_KEY_FILE_PATH)
+        if existing_private_key.read() != private_key:
+            return False
+        existing_ca_certificate = self._container.pull(path=TLS_CA_FILE_PATH)
+        if existing_ca_certificate.read() != ca_certificate:
+            return False
+        return True
+
+    def _set_certificates_secret_in_peer_relation(
+        self,
+        private_key: str,
+        certificate: str,
+        ca_certificate: str,
+    ) -> None:
+        """Set the vault certificate secret in the peer relation.
+
+        Args:
+            private_key: Private key
+            certificate: certificate
+            ca_certificate: CA certificate
+        """
+        if not self._is_peer_relation_created():
+            raise RuntimeError("Peer relation not created")
+        juju_secret_content = {
+            "privatekey": private_key,
+            "certificate": certificate,
+            "cacertificate": ca_certificate,
+        }
+        juju_secret = self.app.add_secret(juju_secret_content, label="vault-certificate")
+        peer_relation = self.model.get_relation(PEER_RELATION_NAME)
+        peer_relation.data[self.app].update({"vault-certificates-secret-id": juju_secret.id})  # type: ignore[union-attr]  # noqa: E501
+        logger.info("Vault certificate secret set in peer relation")
+
+    def _get_certificates_secret_in_peer_relation(
+        self,
+    ) -> Tuple[Optional[str], Optional[str], Optional[str]]:
+        """Get the vault certificate secret from the peer relation.
+
+        Returns:
+            Tuple[Optional[str], Optional[str], Optional[str]]: The private key, certificate and
+                CA certificate.
+        """
+        if not self._is_peer_relation_created():
+            return None, None, None
+        peer_relation = self.model.get_relation(PEER_RELATION_NAME)
+        juju_secret_id = peer_relation.data[peer_relation.app].get(  # type: ignore[union-attr, index]  # noqa: E501
+            "vault-certificates-secret-id"
+        )
+        if not juju_secret_id:
+            return None, None, None
+        try:
+            juju_secret = self.model.get_secret(id=juju_secret_id)
+        except SecretNotFoundError:
+            return None, None, None
+        content = juju_secret.get_content()
+        return content["privatekey"], content["certificate"], content["cacertificate"]
```

Comment:
I think this could be simplified to simply raise in error cases, and have an `except` in the event handler to set the status to waiting, it would get rid of a lot of conditionals.

Reply:
Done, let me know what you think